Edits 1 7 and 1 8 (#27)
bhoffman0/CSAwesome2: _sources/Unit1-Using-Objects-and-Methods/topic-1-8-comments.rst

- @param   Parameter of a method
- @return  Return value for a method

Preconditions and Postconditions
----------------------------------

Many methods in API libraries have **preconditions** and the **postconditions** described in their comments. A **precondition** is a condition that must be true for your method code to work, for example the assumption that the parameters, the data that you give the method to do its job, have values within limits and are not null. The methods could check for these preconditions, but they do not have to. The precondition is what the method expects in order to do its job properly.

A **postcondition** is a condition that is true after running the method. It is what the method promises to do. Postconditions describe the outcome of running the method, for example what is being returned from the method or the changes to the state. These assumptions are very useful to other programmers who want to use that method and get the correct results.

> One precondition that we talked about is that divisors cannot be zero in expressions because that will cause the code to have a runtime error or exception. Many math functions have preconditions about their operands. For example, computing the square root of a negative number is undefined, so the ``Math.sqrt(num)`` Java method which we will learn later will throw an exception (a runtime error) if num is negative. These are described as special cases in the API documentation, for example in https://docs.oracle.com/javase%2F8%2Fdocs%2Fapi%2F%2F/java/lang/Math.html#sqrt-double-. Try this out below.

Reviewer: ```suggestion
One precondition that we talked about is that divisors cannot be zero in expressions because that will cause the code to have a runtime error or exception. Many math functions have preconditions about their operands because not every mathematical operation is defined for every value. For example, computing the square root of a negative number is undefined, so the ``Math.sqrt(num)`` Java method, which we will learn later, will return a special value ``NaN`` which stands for "not a number" if num is negative. But since you can't really do anything useful with ``NaN`` it's better to think of ``sqrt`` as having a precondition that says it only works properly if given a positive argument. These are described as special cases in the API documentation, for example in https://docs.oracle.com/javase%2F8%2Fdocs%2Fapi%2F%2F/java/lang/Math.html#sqrt-double-. Try this out below.
```

|CodingEx| **Coding Exercise**

.. activecode:: math-preconditions
   :language: java
   :autograde: unittest

>    The following code uses the square root method in Java which has a precondition that the number that you give it is not negative.  When you click on run, the compiler will not catch the error, but there will be a run-time error or exception.  Can you fix the value of num so that it does not throw an exception?  What is the precondition for the Math.sqrt method?

Reviewer: ```suggestion
   The following code uses the square root method in Java which has a precondition that the number that you give it is not negative.  When you click on run, the compiler will not catch the error, but it will return the error value ``NaN``.  Can you fix the value of num so that it does not return ``NaN``?  What is the precondition for the Math.sqrt method?
```
